Making a scroller composited generates repaints for two rendering updates.

mattwoodrow opened the pull request:
#### 1e1ce773deb4727ab1fd65317fc61b9592ba44fc
<pre>
Making a scroller composited generates repaints for two rendering updates.
<a href="https://bugs.webkit.org/show_bug.cgi?id=276465">https://bugs.webkit.org/show_bug.cgi?id=276465</a>
&lt;<a href="https://rdar.apple.com/131502835">rdar://131502835</a>&gt;

Reviewed by Simon Fraser.

Creating a newly composited scroller paints the scroller in the following
rendering update, and then repaints it again (as it switches to painting outside
of the directly visible area) on the following rendering update.

One issue is that RenderLayer::updateLayerPositionsAfterLayout computes repaint
rects relative to the nearest composited ancestor, but this runs before
computeCompositingRequirements, so the scroller isn&apos;t considered composited.
Elements inside the scroller get rects relative to the root, and get clipped by
the scroller, so anything not directly visible has an empty rect. On the
successive paint, the scroller is the composited ancestor, so unclipped rects
get generated and new repaints are issued for the change.

The fix for this is to have
RenderLayerScrollableArea::computeHasCompositedScrollableOverflow call
layerStartedCompositingScrollableOverflow (a simplified version of
layerStyleChanged) so that the scroller gets a compositing layer immediately and
updateLayerPositions can compute the right rects initially.

A second issue is that RenderLayerScrollableArea::usesCompositedScrolling()
depends on the presence of RenderLayerBackingStore::m_scrollContainerLayer,
which also isn&apos;t created until compositing layers are flushed (after
updateLayerPositions has finished).

This fix is to check hasCompositedScrollableOverflow() instead, which is the
input to deciding if a scroll container layer will be created in the future.

New test added that creates a composited scroller, and then checks the coverage
rects after the first rendering update. This only has the right rects with the
second, third and fourth issues fixed.

It then mutates an element inside the scroller (but isn&apos;t visible) and checks
that the repaint rects are exactly what changed. Without the first bug fixed,
the repaint rects cover the whole scrollable area.

* LayoutTests/compositing/repaint/scroller-repaints-once-expected.txt: Added.
* LayoutTests/compositing/repaint/scroller-repaints-once.html: Added.
* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/ios/compositing/repaint/scroller-repaints-once-expected.txt: Added.
* LayoutTests/platform/mac-wk1/TestExpectations:
* Source/WebCore/rendering/RenderLayerCompositor.cpp:
(WebCore::RenderLayerCompositor::layerStartedCompositingScrollableOverflow):
* Source/WebCore/rendering/RenderLayerCompositor.h:
* Source/WebCore/rendering/RenderLayerScrollableArea.cpp:
(WebCore::RenderLayerScrollableArea::computeScrollDimensions):
(WebCore::RenderLayerScrollableArea::computeHasCompositedScrollableOverflow):
(WebCore::RenderLayerScrollableArea::usesCompositedScrolling const):
* Source/WebCore/rendering/RenderLayerScrollableArea.h:

Canonical link: <a href="https://commits.webkit.org/281062@main">https://commits.webkit.org/281062@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/bd0911be8e065af7ffae03b03a7959ec6e147566

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/58520 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/37847 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/11004 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/62146 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/8964 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/60649 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/45483 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/9161 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/47375 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/6383 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/60551 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/35426 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/50592 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/28226 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/32176 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/7894 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/7968 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/54127 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/8165 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/63849 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/2433 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/8174 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/54695 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/2441 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/50618 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/54771 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12934 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/2053 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/33676 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/34762 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/35846 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/34507 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/rendering/RenderLayerCompositor.cpp:
```diff
@@ -1972,6 +1972,28 @@ static bool recompositeChangeRequiresChildrenGeometryUpdate(const RenderStyle& o
         || oldStyle.usedTransformStyle3D() != newStyle.usedTransformStyle3D();
 }
 
+void RenderLayerCompositor::layerStartedCompositingScrollableOverflow(RenderLayer& layer)
```
"started" -> "gained"

smfr commented on Source/WebCore/rendering/RenderLayerScrollableArea.cpp:
```diff
@@ -1122,7 +1122,7 @@ void RenderLayerScrollableArea::computeScrollDimensions()
 
     computeScrollOrigin();
-    computeHasCompositedScrollableOverflow(LayoutUpToDate::Yes);
+    computeHasCompositedScrollableOverflow(LayoutUpToDate::Yes, true);
```
Boolean trap. Can the argument be an enum?

mattwoodrow committed:
Making a scroller composited generates repaints for two rendering updates.
https://bugs.webkit.org/show_bug.cgi?id=276465
<rdar://131502835>

Reviewed by Simon Fraser.

Creating a newly composited scroller paints the scroller in the following
rendering update, and then repaints it again (as it switches to painting outside
of the directly visible area) on the following rendering update.

One issue is that RenderLayer::updateLayerPositionsAfterLayout computes repaint
rects relative to the nearest composited ancestor, but this runs before
computeCompositingRequirements, so the scroller isn't considered composited.
Elements inside the scroller get rects relative to the root, and get clipped by
the scroller, so anything not directly visible has an empty rect. On the
successive paint, the scroller is the composited ancestor, so unclipped rects
get generated and new repaints are issued for the change.

The fix for this is to have
RenderLayerScrollableArea::computeHasCompositedScrollableOverflow call
layerStartedCompositingScrollableOverflow (a simplified version of
layerStyleChanged) so that the scroller gets a compositing layer immediately and
updateLayerPositions can compute the right rects initially.

A second issue is that RenderLayerScrollableArea::usesCompositedScrolling()
depends on the presence of RenderLayerBackingStore::m_scrollContainerLayer,
which also isn't created until compositing layers are flushed (after
updateLayerPositions has finished).

This fix is to check hasCompositedScrollableOverflow() instead, which is the
input to deciding if a scroll container layer will be created in the future.

New test added that creates a composited scroller, and then checks the coverage
rects after the first rendering update. This only has the right rects with the
second, third and fourth issues fixed.

It then mutates an element inside the scroller (but isn't visible) and checks
that the repaint rects are exactly what changed. Without the first bug fixed,
the repaint rects cover the whole scrollable area.

* LayoutTests/compositing/repaint/scroller-repaints-once-expected.txt: Added.
* LayoutTests/compositing/repaint/scroller-repaints-once.html: Added.
* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/ios/compositing/repaint/scroller-repaints-once-expected.txt: Added.
* LayoutTests/platform/mac-wk1/TestExpectations:
* Source/WebCore/rendering/RenderLayerCompositor.cpp:
(WebCore::RenderLayerCompositor::layerStartedCompositingScrollableOverflow):
* Source/WebCore/rendering/RenderLayerCompositor.h:
* Source/WebCore/rendering/RenderLayerScrollableArea.cpp:
(WebCore::RenderLayerScrollableArea::computeScrollDimensions):
(WebCore::RenderLayerScrollableArea::computeHasCompositedScrollableOverflow):
(WebCore::RenderLayerScrollableArea::usesCompositedScrolling const):
* Source/WebCore/rendering/RenderLayerScrollableArea.h:

Canonical link: https://commits.webkit.org/281062@main
